File: 백준/Silver/2167. 2차원 배열의 합/2차원 배열의 합.py
from sys import stdin

n, m = map(int, stdin.readline().split())
arrays = [[0]*(m+2)] + [[0] + list(map(int, stdin.readline().split())) + [0] for _ in range(n)]
for i in range(1, n+1):
    for j in range(1, m+1):
        arrays[i][j] += arrays[i][j-1]
    for j in range(1, m+1):
        arrays[i][j] += arrays[i-1][j]
        

k = int(stdin.readline())
for _ in range(k):
    i, j, x, y = map(int, stdin.readline().split())
    print(arrays[x][y] - arrays[i-1][y] - arrays[x][j-1] + arrays[i-1][j-1])

File: 백준/Silver/2167. 2차원 배열의 합/2차원 배열의 합.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken()), m = Integer.parseInt(st.nextToken());
        int[][] arrays = new int[n + 1][m + 2];
        for (int i = 1; i < n + 1; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 1; st.hasMoreTokens(); j++) {
                arrays[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        for (int i = 1; i < n + 1; i++) {
            for (int j = 1; j < m + 1; j++) {
                arrays[i][j] += arrays[i][j - 1];
            }
            for (int j = 1; j < m + 1; j++) {
                arrays[i][j] += arrays[i - 1][j];
            }
        }

        int k = Integer.parseInt(br.readLine());
        for (int t = 0; t < k; t++) {
            st = new StringTokenizer(br.readLine());
            int i = Integer.parseInt(st.nextToken()), j = Integer.parseInt(st.nextToken()), x = Integer.parseInt(st.nextToken()), y = Integer.parseInt(st.nextToken());
            System.out.println(arrays[x][y] - arrays[i - 1][y] - arrays[x][j - 1] + arrays[i - 1][j - 1]);
        }
    }
}